### PR TITLE
Replumbing: Add constructor of libcrypto internal method  structures

### PR DESCRIPTION
**Pull request by levitte**
This queries the provider for its available functionality (unless a
matching method structured is already cached, in which case that's
used instead), and creates method structure with the help of a passed
constructor.  The result is cached if the provider allows it (or if
caching is forced).

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
Note that this PR is made on top of #8286 and #8287, so there are a number of commits that don't really belong here.  To avoid confusion, I'd like for the other two to be merged before commentary here, but I thought I'd provide this to show how the other two are going to be used.

File: crypto/provider_core.c
Comment (paulidale) [inline crypto/provider_core.c]: Not related to this PR, but we might want to macro wrap this pattern before it becomes too pervasive.

File: doc/internal/man3/ossl_method_construct.pod
Comment (paulidale) [inline doc/internal/man3/ossl_method_construct.pod]: s/in/of/

File: crypto/provider_core.c
Comment (levitte) [reply crypto/provider_core.c]: PR welcome! :wink:

File: doc/internal/man3/ossl_method_construct.pod
Comment (paulidale) [inline doc/internal/man3/ossl_method_construct.pod]: @t-j-h thought keeping a reference was a bad idea.

**Review by paulidale — APPROVED** [review]
Some of the comments in #8287 likely apply to this PR instead.

One wording nit and one that might be a mis-remember on my part.

File: doc/internal/man3/ossl_method_construct.pod
Comment (levitte) [reply doc/internal/man3/ossl_method_construct.pod]: I know where he's coming from, but am worried about crashes because of pointers that suddenly lead nowhere.  That seems like a bad idea as well.  So if underlying providers shall be allowed to do a disappearing act, we need to figure out a workable mechanism to detect that.

Comment (levitte) [reply doc/internal/man3/ossl_method_construct.pod]: Note that this isn't necessarily something that needs to block this PR, as long as we don't forget.

**Comment by levitte** [comment]
#8287 is now merged.  I've rebased this and squashed all fixups

**Comment by levitte** [comment]
I know that @paulidale approved this a while ago, but I think we do need a final review now that it doesn't depend on any other PR any more.

File: crypto/core_fetch.c
Comment (paulidale) [inline crypto/core_fetch.c]: Is it worth using a cast as well?  C++ doesn’t always like don’t doing this implicitly.

File: crypto/core_fetch.c
Comment (paulidale) [inline crypto/core_fetch.c]: A nice mix or names using underscores and not...

File: doc/internal/man3/ossl_method_construct.pod
Comment (paulidale) [inline doc/internal/man3/ossl_method_construct.pod]: That want~s~ to

File: doc/internal/man3/ossl_method_construct.pod
Comment (paulidale) [inline doc/internal/man3/ossl_method_construct.pod]: The _just_ isn’t required.

File: doc/internal/man3/ossl_method_construct.pod
Comment (paulidale) [inline doc/internal/man3/ossl_method_construct.pod]: Temporarily, there is a missing ‘i’

File: doc/internal/man3/ossl_method_construct.pod
Comment (paulidale) [inline doc/internal/man3/ossl_method_construct.pod]: Term instead of time?

File: doc/internal/man3/ossl_method_construct.pod
Comment (paulidale) [inline doc/internal/man3/ossl_method_construct.pod]: Really?

**Review by paulidale — APPROVED** [review]
A couple of documentation nit.

File: crypto/core_fetch.c
Comment (levitte) [reply crypto/core_fetch.c]: This isn't C++

**Comment by levitte** [comment]
Final review, please?

File: doc/internal/man3/ossl_method_construct.pod
Comment (levitte) [reply doc/internal/man3/ossl_method_construct.pod]: Fixed

File: doc/internal/man3/ossl_method_construct.pod
Comment (levitte) [reply doc/internal/man3/ossl_method_construct.pod]: Fixed

File: doc/internal/man3/ossl_method_construct.pod
Comment (levitte) [reply doc/internal/man3/ossl_method_construct.pod]: (I'll never be able to remember...)

Fixed

File: doc/internal/man3/ossl_method_construct.pod
Comment (levitte) [reply doc/internal/man3/ossl_method_construct.pod]: Sure.

File: doc/internal/man3/ossl_method_construct.pod
Comment (levitte) [reply doc/internal/man3/ossl_method_construct.pod]: Uhmmmm, no

**Comment by levitte** [comment]
Merged.

9e11fe0d85c7d8bd2b77076c8b2e93433091e765 Replumbing: Add constructor of libcrypto internal method structures
099bd33920e775eb75f4daee5f09b24f17bc136d Replumbing: Add support for the provider query_operation function
85e2417c0d81cfe97586b74c410ef37595aea72d Replumbing: Add an OSSL_PROVIDER iterator with callback

**Pull request closed without merging**
